src/els_background_job.erl

```diff
@@ -0,0 +1,190 @@
+  ok.
+
+%% @doc Start the server responsible for a background job
+%%
+%% To be used by the supervisor
+-spec start_link(config()) -> {ok, pid()}.
+start_link(Config) ->
+  gen_server:start_link(?MODULE, Config, []).
+
+%%==============================================================================
+%% Callbacks for gen_server
+%%==============================================================================
+-spec init(config()) -> {ok, state()}.
+init(#{entries := Entries, title := Title} = Config) ->
+  lager:info("Background job started. [pid=~p] [title=~p]", [self(), Title]),
+  %% Ensure the terminate function is called on shutdown, allowing the
+  %% job to clean up.
+  process_flag(trap_exit, true),
+  ProgressEnabled = els_work_done_progress:is_supported(),
+  Total = length(Entries),
+  Step = step(Total),
+  Token = els_work_done_progress:send_create_request(),
+  notify_begin(Token, Title, Total, ProgressEnabled),
+  self() ! exec,
+  {ok, #{ config => Config
+        , progress_enabled => ProgressEnabled
+        , token => Token
+        , current => 0
+        , step => Step
+        , total => Total
+        }}.
+
+-spec handle_call(any(), {pid(), any()}, state()) ->
+  {noreply, state()}.
+handle_call(_Request, _From, State) ->
+  {noreply, State}.
+
+-spec handle_cast(any(), any()) ->
+  {noreply, state()}.
+handle_cast(_Request, State) ->
+  {noreply, State}.
+
+-spec handle_info(any(), any()) ->
+  {noreply, state()}.
+handle_info(exec, State) ->
+  #{ config := #{ entries := Entries, task := Task} = Config
+   , progress_enabled := ProgressEnabled
+   , token := Token
+   , current := Current
+   , step := Step
+   , total := Total
+   } = State,
+  case Entries of
+    [] ->
+      notify_end(Token, Total, ProgressEnabled),
+      {stop, normal, State};
+    [Entry|Rest] ->
+      Task(Entry),
+      notify_report(Token, Current, Step, Total, ProgressEnabled),
+      self() ! exec,
+      {noreply, State#{ config => Config#{ entries => Rest }
+                      , current => Current + 1}}
+  end;
+handle_info(_Request, State) ->
+  {noreply, State}.
+
+-spec terminate(any(), state()) -> ok.
+terminate(normal, #{config := Config}) ->
+  lager:info("Background job completed. [pid=~p]", [self()]),
+  OnComplete = maps:get(on_complete, Config, noop()),
+  OnComplete(),
+  ok;
+terminate(Reason, #{config := Config}) ->
+  lager:warning( "Background job aborted. [reason=~p] [pid=~p]"
+               , [Reason, self()]),
+  OnError = maps:get(on_error, Config, noop()),
+  OnError(),
+  ok.
+
+%%==============================================================================
+%% Internal functions
+%%==============================================================================
+-spec step(pos_integer()) -> pos_integer().
+step(0) -> 0;
+step(N) -> 100 / N.
+
+-spec progress_msg(non_neg_integer(), pos_integer()) -> binary().
+progress_msg(Current, Total) ->
+  list_to_binary(io_lib:format("~p / ~p", [Current, Total])).
```

Comment:
I think we need to add the "No Call" rule like @alanz suggested to avoid the usage of this functions.

Reply:
Not sure if that's too restrictive, but feel free to add it and then we can decide if keeping or not. Fixing this occurrence in the meantime.

Reply:
I think we should also add this to the guidelines.

Reply:
In retrospection, since this seems to be the only change I would make, I would prefer to merge it as-is and fix once we introduce the elvis restriction (so we can try it).